engine/version/version_controller.hpp

```diff
@@ -70,6 +70,73 @@ class SnapshotList {
 // The snapshots include temporal snapshots that cached by each access thread of
 // kvdk instance, and a global snapshot list that actively created by user
 class VersionController {
+ public:
+  // LocalSnapshotHolder is used by kvdk functions such as Get(), SGet() and
+  // HashGet() internally to guarantee lockless reads will always read out valid
+  // data. LocalSnapshotHolder is thread_local, and one thread can hold atmost
+  // one at same time.
+  class LocalSnapshotHolder {
+    VersionController* owner_{nullptr};
+    TimeStampType ts_{};
+
+   public:
+    LocalSnapshotHolder(VersionController* o) : owner_{o} {
+      owner_->HoldLocalSnapshot();
+      ts_ = owner_->version_thread_cache_[access_thread.id]
+                .holding_snapshot.timestamp;
+    };
+    LocalSnapshotHolder(LocalSnapshotHolder const&) = delete;
+    LocalSnapshotHolder& operator=(LocalSnapshotHolder const&) = delete;
+    LocalSnapshotHolder(LocalSnapshotHolder&& other) {
+      *this = std::move(other);
+    }
+    LocalSnapshotHolder& operator=(LocalSnapshotHolder&& other) {
+      kvdk_assert(owner_ == nullptr, "");
+      kvdk_assert(other.owner_ != nullptr, "");
+      std::swap(owner_, other.owner_);
+      std::swap(ts_, other.ts_);
+      return *this;
+    }
+    ~LocalSnapshotHolder() {
+      if (owner_ != nullptr) {
+        owner_->ReleaseLocalSnapshot();
+      }
+    }
+    TimeStampType Timestamp() { return ts_; }
+  };
+
+  // GlobalSnapshotHolder is hold internally by iterators.
+  // Create GlobalSnapshotHolder is more costly then InteralToken.
+  class GlobalSnapshotHolder {
+    VersionController* owner_{nullptr};
+    SnapshotImpl* snap_{nullptr};
+
+   public:
+    GlobalSnapshotHolder(VersionController* o) : owner_{o} {
+      snap_ = owner_->NewGlobalSnapshot();
+    };
+    GlobalSnapshotHolder(GlobalSnapshotHolder const&) = delete;
+    GlobalSnapshotHolder& operator=(GlobalSnapshotHolder const&) = delete;
+    GlobalSnapshotHolder(GlobalSnapshotHolder&& other) {
+      *this = std::move(other);
+    }
+    GlobalSnapshotHolder& operator=(GlobalSnapshotHolder&& other) {
```

Comment:
where need to `=`?

Reply:
Transfer ownership. Move constructor and move assignment should be defined together. It's suggested by cpp core guidelines.

Reply:
if we don't use `=`, abandon？